csm_web/frontend/src/components/course/ManageLabelsModal.tsx

```diff
@@ -0,0 +1,207 @@
+    setCourseLabels(newLabels);
+  };
+
+  const handleNewRow = () => {
+    const course = courseLabels[0].course;
+    setCourseLabels([
+      ...courseLabels,
+      { id: newLabelKey, course: course, sections: [], name: "", description: "", showPopup: false }
+    ]);
+    setNewLabelKey(newLabelKey - 1);
+  };
+
+  function handleSubmit(event: React.ChangeEvent<HTMLFormElement>) {
+    event.preventDefault();
+    const currIDs = courseLabels.map((label: Label) => label.id);
+    const prevIDs = currCourseLabels.map((label: Label) => label.id);
+    const deletedIDs = [];
+    for (const id of prevIDs) {
+      if (!currIDs.includes(id)) {
+        deletedIDs.push(id);
+      }
+    }
+    const put_value = {
+      labels: courseLabels,
+      deletedLabelIds: deletedIDs
+    };
+    fetchWithMethod(`/courses/${courseId}/labels`, HTTP_METHODS.PUT, put_value).then(() => {
+      closeModal();
+      reloadSections();
+    });
+  }
+
+  return (
+    <Modal closeModal={closeModal}>
+      <form className="manage-labels-form" onSubmit={handleSubmit}>
+        <div>
+          <div className="manage-labels-form-header">Labels for {title}</div>
+          <div className="manage-labels-modal">
+            <ul>
+              {courseLabels.map((label: Label) => (
+                <EditLabelRow
+                  key={label.id}
+                  label={label}
+                  onChangeRow={(name: string, description: string, showPopup: boolean) =>
+                    editRow(label.id, name, description, showPopup)
+                  }
+                  removeLabel={() => removeLabel(label.id)}
+                />
+              ))}
+            </ul>
+          </div>
+          <div className="manage-label-modal-buttons">
+            <input className="affinity-btn" type="button" onClick={handleNewRow} value="Add New Label" />
+            <input className="affinity-btn" type="submit" value="Save Labels" />
+          </div>
+        </div>
+      </form>
+    </Modal>
+  );
+};
+
+interface EditLabelRowProps {
+  label: Label;
+  onChangeRow: (name: string, description: string, showPopup: boolean) => void;
+  removeLabel: (id: number) => void;
+}
+
+export const EditLabelRow = ({ label, onChangeRow, removeLabel }: EditLabelRowProps) => {
+  const [editingLabels, setEditingLabels] = useState<boolean>(false);
+
+  const toggleEditingLabels = (e: React.MouseEvent<HTMLOrSVGElement, MouseEvent>) => {
+    e.stopPropagation();
+    setEditingLabels(!editingLabels);
+  };
+
+  const editValue = (type: string, value: any) => {
+    if (type == "name") {
+      onChangeRow(value, label.description, label.showPopup);
+    } else if (type == "description") {
+      onChangeRow(label.name, value, label.showPopup);
+    } else if (type == "showpopup") {
+      onChangeRow(label.name, label.description, value);
+    } else {
+      // should not get here
+      console.error(`Unknown attribute ${type}`);
+    }
```

Comment:
This is probably better replaced with a `switch` statement, and the types can also be changed to enums, as it seems like the type string is only used for identification purposes, to distinguish between which fields are being edited.